File: src/main/java/uk/gov/justice/hmpps/casenotes/dto/CaseNoteType.java
package uk.gov.justice.hmpps.casenotes.dto;

import com.fasterxml.jackson.annotation.JsonInclude;
import com.fasterxml.jackson.annotation.JsonPropertyOrder;
import io.swagger.annotations.ApiModel;
import io.swagger.annotations.ApiModelProperty;
import lombok.AllArgsConstructor;
import lombok.Builder;
import lombok.Data;
import lombok.EqualsAndHashCode;
import lombok.NoArgsConstructor;
import lombok.ToString;

import java.util.ArrayList;
import java.util.List;

@ApiModel(description = "Case Note Type")
@JsonInclude(JsonInclude.Include.NON_NULL)
@AllArgsConstructor
@NoArgsConstructor
@EqualsAndHashCode
@JsonPropertyOrder({"code", "description", "activeFlag", "subCodes"})
@Data
@ToString
@Builder(toBuilder = true)
public class CaseNoteType implements Comparable<CaseNoteType>{

    @ApiModelProperty(required = true, value = "Case Note Code", position = 1, example = "OBSERVE")
    private String code;

    @ApiModelProperty(required = true, value = "Case Note description.", position = 2, example = "Observations")
    private String description;

    @ApiModelProperty(required = true, value = "Active indicator flag.", example = "Y", allowableValues = "Y,N", position = 3)
    @Builder.Default
    private String activeFlag = "Y";

    @ApiModelProperty(value = "Source of Case Note Type, legacy case note are null", example = "OCNS", position = 4)
    private String source;

    @ApiModelProperty(value = "Indicates the type of note is sensitive", example = "true", position = 5)
    private boolean sensitive;

    @ApiModelProperty(value = "List of case note sub types", position = 6, allowEmptyValue = true)
    @Builder.Default
    private List<CaseNoteType> subCodes = new ArrayList<>();

    @Override
    public int compareTo(final CaseNoteType type) {
        return getDescription().compareToIgnoreCase(type.getDescription());
    }
}

File: src/main/java/uk/gov/justice/hmpps/casenotes/types/CaseNoteType.kt
package uk.gov.justice.hmpps.casenotes.types

import io.swagger.v3.oas.annotations.media.Schema

interface CodedValue : Comparable<CodedValue> {
  val code: String
  val description: String
  override fun compareTo(other: CodedValue): Int {
    val dif = description.compareTo(other.description, ignoreCase = true)
    return if (dif == 0) code.compareTo(other.code) else dif
  }
}

data class CaseNoteType(
  override val code: String,
  override val description: String,
  @Schema(description = "List of case note sub types")
  val subCodes: List<CaseNoteSubType> = listOf(),
) : CodedValue

data class CaseNoteSubType(
  @Schema(required = true, description = "Case Note Code", example = "OBSERVE")
  override val code: String,

  @Schema(required = true, description = "Case Note description.", example = "Observations")
  override val description: String,

  @Schema(required = true, description = "Indicates if the type is active or not")
  val active: Boolean,

  @Schema(description = "Indicates the type of note is sensitive", example = "true")
  val sensitive: Boolean,

  @Schema(
    description = "Indicates the type of note can only be created by a sub-set of users (e.g. POMs)",
    example = "true",
  )
  val restrictedUse: Boolean,

  @Schema(description = "Shows the actors that can select this case note type")
  val selectableBy: List<SelectableBy>,
) : CodedValue {

  @Deprecated("to be replaced with 'active' boolean")
  val activeFlag: ActiveYn = if (active) ActiveYn.Y else ActiveYn.N
}

enum class ActiveYn {
  Y,
  N,
}
